### PR TITLE
WIP: Update PodStatus with multiple IP's for Dual Stack

### PR DESCRIPTION
**Pull request by rpothier**
**What this PR does / why we need it**:
This PR adds the foundation for multiple IP's in the PodStatus API.
There is no functional changes in this PR, but as only one IP is handled.
The KEP for this change is https://github.com/kubernetes/community/pull/2254
The API review for this is https://github.com/kubernetes-sigs/architecture-tracking/issues/52

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
The API changes have not yet been approved, but adding this review
as a reference in the discussion.

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by leblancd** [comment]
@caesarxuchao , @eparis - Would it be okay to add an "okay-to-test" this change, even though it's still a WIP? We're thinking that this change shouldn't affect current functionality, but we'd like to check the CI test results.

File: pkg/apis/core/v1/conversion.go [hunk 1]
```diff
@@ -327,6 +329,35 @@ func Convert_v1_Pod_To_core_Pod(in *v1.Pod, out *core.Pod, s conversion.Scope) e
 	return nil
 }
 
+func Convert_v1_PodStatus_To_core_PodStatus(in *v1.PodStatus, out *core.PodStatus, s conversion.Scope) error {
+	autoConvert_v1_PodStatus_To_core_PodStatus(in, out, s)
+	// autoconvert sets the core PodIPs to the v1 PodIPs
```
Comment (leblancd) [inline pkg/apis/core/v1/conversion.go]: Chatted about this with you earlier, but the autoconvert function is going to overwrite the existing core podIPs, so you'll lose that history. Probably need to save that array before calling autoconvert.

File: pkg/apis/core/v1/conversion.go [hunk 2]
```diff
@@ -327,6 +329,35 @@ func Convert_v1_Pod_To_core_Pod(in *v1.Pod, out *core.Pod, s conversion.Scope) e
...
+	// but does not handle the singular PodIP
+
+	if len(in.PodIPs) > 0 && in.PodIP != "" {
+		if in.PodIPs[0].IP != in.PodIP {
+			return fmt.Errorf("Invalid PodIP")
+		}
+	}
+
+	if len(out.PodIPs) > 0 {
```
Comment (leblancd) [inline pkg/apis/core/v1/conversion.go]: Looks like this doesn't handle the case where in.PodIPs[] is set, but in.PodIP is not set (e.g. an update operation). Maybe what we need would be something like this:
```
    // Save existing PodIPs array before autoconvert overwrites it
    savedPodIPs = out.PodIPs
    autoConvert_v1_PodStatus_To_core_PodStatus(in, out, s)

    if len(in.PodIPs) == 0 {
        if in.PodIP != "" {
            // Only PodIP is provided
            // Restore saved PodIPs
            out.PodIPs = savedPodIPs
            out.PodIPs[0] = in.PodIP
            // DELETE ANY REDUNDANT ENTRIES IN out.PodIPs[] HERE
        } else {
            // Nothing to do
            return nil
        }
    } else if in.PodIP == "" {
        // Only PodIPs[] provided
        // DELETE ANY REDUNDANT ENTRIES IN out.PodIPs[] HERE
    } else {
        // Both PodIP and PodIPs[] provided
        if in.PodIPs[0].IP != in.PodIP {
            return fmt.Errorf("Invalid PodIP")
        }
    }
```

File: pkg/registry/core/service/storage/rest.go [hunk 3]
```diff
@@ -521,8 +521,8 @@ func isValidAddress(ctx context.Context, addr *api.EndpointAddress, pods rest.Ge
 	if pod == nil {
 		return fmt.Errorf("pod is missing, skipping (%s/%s)", addr.TargetRef.Namespace, addr.TargetRef.Name)
 	}
-	if pod.Status.PodIP != addr.IP {
-		return fmt.Errorf("pod ip doesn't match endpoint ip, skipping: %s vs %s (%s/%s)", pod.Status.PodIP, addr.IP, addr.TargetRef.Namespace, addr.TargetRef.Name)
+	if pod.Status.PodIPs[0].IP != addr.IP {
```
Comment (leblancd) [inline pkg/registry/core/service/storage/rest.go]: Do we need a sanity length check on PodIPs[] before referencing the [0] entry?

**Review by leblancd — CHANGES_REQUESTED** [review]

File: pkg/registry/core/service/storage/rest.go (hunk 3 above)
Comment (rpothier) [reply pkg/registry/core/service/storage/rest.go]: yes, thanks

File: pkg/apis/core/v1/conversion.go (hunk 2 above)
Comment (rpothier) [reply pkg/apis/core/v1/conversion.go]: Yea, for this case the podIPs that is passed with the versioned, will be set in the internal version through the
autoconvert function. If the previous value is not passed in, then there are no previous values to delete.
The previous values are not passed in from what I  can see.

**Pull request closed without merging**